tests/ui/async-await/pin-ergonomics/project-pattern-match.rs

```diff
@@ -0,0 +1,192 @@
+//@ revisions: pin_project normal
+//@ edition:2024
+//@[pin_project] check-pass
+#![cfg_attr(pin_project, feature(pin_ergonomics))]
+#![feature(if_let_guard)]
+#![allow(incomplete_features)]
+
+use std::pin::Pin;
+
+// This test verifies that a `&pin mut Foo` can be projected to a pinned
+// reference `&pin mut T` of a `?Unpin` field , and can be projected to
+// an unpinned reference `&mut U` of an `Unpin` field .
+
+struct Foo<T, U> {
+    x: T,
+    y: U,
+}
+
+struct Bar<T, U>(T, U);
+
+enum Baz<T, U> {
+    Foo(T, U),
+    Bar { x: T, y: U },
+}
+
+trait IsPinMut {}
+trait IsPinConst {}
+trait IsRefMut {}
+trait IsRef {}
+impl<T: ?Sized> IsPinMut for Pin<&mut T> {}
+impl<T: ?Sized> IsPinConst for Pin<&T> {}
+impl<T: ?Sized> IsRefMut for &mut T {}
+impl<T: ?Sized> IsRef for &T {}
+
+fn assert_pin_mut<T: IsPinMut>(_: T) {}
+fn assert_pin_const<T: IsPinConst>(_: T) {}
+fn assert_ref_mut<T: IsRefMut>(_: T) {}
+fn assert_ref<T: IsRef>(_: T) {}
+
+fn foo_mut<T, U: Unpin>(foo: Pin<&mut Foo<T, U>>) {
+    let Foo { x, y } = foo;
+    //[normal]~^ ERROR mismatched types
+    assert_pin_mut(x);
+    assert_ref_mut(y);
+}
```

Comment:
Something else we'll probably want to do in follow-up are explicit `ref pin mut` / `ref pin const` binding modes.

E.g., we can write today:

```rust
fn f<T>(&mut [ref mut x]: &mut [T; 1]) -> &mut T { x }
```

Probably we'd similarly want something along the lines of:

```rust
fn f<T>(&pin mut [ref pin mut x]: &pin mut [T; 1]) -> &pin mut T { x }
```